### PR TITLE
Cleanup the concept of orderedGroups. Just use a single way of access…

### PR DESCRIPTION
**Pull request by baldersheim**
…ing the groups.

Simplify testing by introducing a GroupList to contain all acces to groups and nodes.

@bjorncs PR

### CODE REVIEW

**Comment by baldersheim** [comment]
Just hold the review. Not ready yet.

**Comment by baldersheim** [comment]
@bjorncs Now it is ready for review

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchGroupsImpl.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    @Override
+    public boolean isPartialGroupCoverageSufficient(Collection<Node> nodes) {
+        if (size() == 1)
+            return true;
+        long activeDocuments = nodes.stream().mapToLong(Node::getActiveDocuments).sum();
+        return isGroupCoverageSufficient(activeDocuments, medianDocumentsPerGroup());
+    }
+
+    public boolean isGroupCoverageSufficient(long activeDocuments, long medianDocuments) {
+        double documentCoverage = 100.0 * (double) activeDocuments / medianDocuments;
```
Comment (bjorncs) [inline container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchGroupsImpl.java:33]: Division on zero possible if `isEmpty()` returns `true`?

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchGroupsImpl.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    public long medianDocumentsPerGroup() {
+        if (isEmpty()) return 0;
+        var activeDocuments = groups().stream().map(Group::activeDocuments).collect(Collectors.toList());
+        return (long) Quantiles.median().compute(activeDocuments);
+    }
+
+
+    public static SearchGroupsImpl buildGroupListForTest(int numGroups, int nodesPerGroup, double minActivedocsPercentage) {
```
Comment (bjorncs) [inline container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchGroupsImpl.java:44]: Consider moving to a helper in `src/test/java`

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchCluster.java [hunk 3]
```diff
@@ -109,27 +95,22 @@ private static List<Node> toNodes(DispatchNodesConfig nodesConfig) {
...
-    public List<Group> orderedGroups() { return orderedGroups; }
-
-    /** Returns the n'th (zero-indexed) group in the cluster if possible */
-    public Optional<Group> group(int n) {
-        if (orderedGroups().size() > n) {
-            return Optional.of(orderedGroups().get(n));
-        } else {
-            return Optional.empty();
+    private static SearchGroupsImpl toGroups(Collection<Node> nodes, double minActivedocsPercentage) {
+        ImmutableMap.Builder<Integer, Group> groupsBuilder = new ImmutableMap.Builder<>();
```
Comment (bjorncs) [inline container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchCluster.java:99]: Consider using a `HashMap` + `Map.copyOf()`

**Review by bjorncs — COMMENTED** [review]
See comments

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchCluster.java (hunk 3 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchCluster.java:99]: Fixed

File: container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchGroupsImpl.java (hunk 2 above)
Comment (baldersheim) [reply container-search/src/main/java/com/yahoo/search/dispatch/searchcluster/SearchGroupsImpl.java]: Good point.

**Review by bjorncs — APPROVED** [review]

**Pull request merged**